File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/FactionHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.long
import com.zeroday.handler.reqStr
import com.zeroday.handler.str
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull

class FactionHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.FACTION_CREATE,
        RequestTypes.FACTION_JOIN,
        RequestTypes.FACTION_LEAVE,
        RequestTypes.FACTION_INFO,
        RequestTypes.FACTION_LIST,
        RequestTypes.FACTION_DONATE,
        RequestTypes.FACTION_UPGRADE
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.FACTION_CREATE -> create(ctx, playerId, payload)
            RequestTypes.FACTION_JOIN -> join(ctx, playerId, payload)
            RequestTypes.FACTION_LEAVE -> leave(ctx, playerId)
            RequestTypes.FACTION_INFO -> info(ctx, playerId)
            RequestTypes.FACTION_LIST -> list(ctx)
            RequestTypes.FACTION_DONATE -> donate(ctx, playerId, payload)
            RequestTypes.FACTION_UPGRADE -> upgrade(ctx, playerId, payload)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported faction request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun create(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val name = payload.str("name")
            ?: return HandlerResult.Err(ServerError.fromMessage("Faction name required", ErrorCategory.VALIDATION))
        val tag = payload.str("tag")
            ?: return HandlerResult.Err(ServerError.fromMessage("Faction tag required", ErrorCategory.VALIDATION))
        val desc = payload.str("description") ?: ""
        return ctx.services.factionService.createFaction(playerId, name, tag, desc).fold(
            onSuccess = { faction -> HandlerResult.Ok(ResponseTypes.FACTION_CREATED, mapOf("faction" to faction)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not create faction", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun join(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val factionId = payload.str("faction_id")
            ?: return HandlerResult.Err(ServerError.fromMessage("Faction ID required", ErrorCategory.VALIDATION))
        return ctx.services.factionService.joinFaction(playerId, factionId).fold(
            onSuccess = { f -> HandlerResult.Ok(ResponseTypes.FACTION_JOINED, mapOf("faction" to f)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not join faction", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun leave(ctx: HandlerContext, playerId: String): HandlerResult =
        ctx.services.factionService.leaveFaction(playerId).fold(
            onSuccess = { HandlerResult.Ok(ResponseTypes.FACTION_LEFT, mapOf("message" to "Left faction")) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not leave faction", ErrorCategory.INTERNAL)) }
        )

    private suspend fun info(ctx: HandlerContext, playerId: String): HandlerResult =
        ctx.services.factionService.getFaction(playerId).fold(
            onSuccess = { f -> HandlerResult.Ok(ResponseTypes.FACTION_INFO, mapOf("faction" to f)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "No faction", ErrorCategory.NOT_FOUND)) }
        )

    private suspend fun list(ctx: HandlerContext): HandlerResult {
        val factions = ctx.services.factionService.listFactions()
        return HandlerResult.Ok(ResponseTypes.FACTION_LIST, mapOf("factions" to factions))
    }

    private suspend fun donate(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val cpu = payload.long("cpu")
        val ram = payload.long("ram")
        val bw = payload.long("bandwidth")
        val credits = payload.long("credits")
        return ctx.services.factionService.donateToMainframe(playerId, cpu, ram, bw, credits).fold(
            onSuccess = { mf -> HandlerResult.Ok(ResponseTypes.FACTION_DONATED, mapOf("mainframe" to mf)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Donation failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun upgrade(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val upgradeId = payload.str("upgrade_id")
            ?: return HandlerResult.Err(ServerError.fromMessage("Upgrade ID required", ErrorCategory.VALIDATION))
        return ctx.services.factionService.upgradeMainframe(playerId, upgradeId).fold(
            onSuccess = { mf -> HandlerResult.Ok(ResponseTypes.FACTION_UPGRADED, mapOf("mainframe" to mf)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Upgrade failed", ErrorCategory.INTERNAL)) }
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/FactionHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class FactionHandler : IHandler
{
    private readonly FactionService _factionService;
    private readonly IPlayerService _playerService;

    public string MessageType => "faction_create";

    public FactionHandler(FactionService factionService, IPlayerService playerService)
    {
        _factionService = factionService;
        _playerService = playerService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var name = payload?.Str("name");
        var tag = payload?.Str("tag");
        var description = payload?.Str("description") ?? "";

        if (string.IsNullOrEmpty(name))
            return new MessageResult("error", connectionId, new { message = "Faction name required" });
        if (string.IsNullOrEmpty(tag))
            return new MessageResult("error", connectionId, new { message = "Faction tag required" });

        try
        {
            var faction = await _factionService.CreateFaction(playerId, name, tag, description);
            return new MessageResult("faction_created", connectionId, new { faction });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleJoin(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var factionId = payload?.Str("faction_id");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(factionId))
            return new MessageResult("error", connectionId, new { message = "Faction ID required" });

        try
        {
            var faction = await _factionService.JoinFaction(playerId, factionId);
            return new MessageResult("faction_joined", connectionId, new { faction });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleLeave(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            await _factionService.LeaveFaction(playerId);
            return new MessageResult("faction_left", connectionId, new { message = "Left faction" });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleInfo(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var faction = await _factionService.GetFaction(playerId);
            return new MessageResult("faction_info", connectionId, new { faction });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleList(string connectionId, JsonElement? payload, string? requestId)
    {
        var factions = await _factionService.ListFactions();
        return new MessageResult("faction_list", connectionId, new { factions });
    }

    public async Task<IActionResult> HandleDonate(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        var cpu = payload?.Long("cpu") ?? 0L;
        var ram = payload?.Long("ram") ?? 0L;
        var bandwidth = payload?.Long("bandwidth") ?? 0L;
        var credits = payload?.Long("credits") ?? 0L;

        try
        {
            var mainframe = await _factionService.DonateToMainframe(playerId, cpu, ram, bandwidth, credits);
            return new MessageResult("faction_donated", connectionId, new { mainframe });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleUpgrade(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var upgradeId = payload?.Str("upgrade_id");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(upgradeId))
            return new MessageResult("error", connectionId, new { message = "Upgrade ID required" });

        try
        {
            var mainframe = await _factionService.UpgradeMainframe(playerId, upgradeId);
            return new MessageResult("faction_upgraded", connectionId, new { mainframe });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }
}
